packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,295 @@
+type instance = any;
+
+type request = (instance: instance) => Promise<any>;
+
+export type poolOptions = {
+	spawn: () => Promise<instance>;
+	reap?: (instance: instance) => void;
+	fatal?: (instance: instance, error: any) => any;
+	maxRequests?: number;
+	maxJobs?: number;
+};
+
+let childCount = 0;
+
+/**
+ * Tracks stats of instances in a pool.
+ * @private
+ */
+class PoolInfo {
+	id = childCount++; // Unique ID for debugging purposes.
+	requests = 0; // Total requests processed.
+	started = Date.now(); // Time spawned.
+	active = false; // Whether instance is considered active.
+}
+
+/**
+ * Spawns new instances if the pool is not full.
+ * Returns a list of new instances.
+ * @param pool the pool object to work on
+ * @private
+ */
+const spawn = async (pool: Pool) => {
+	const newInstances = new Set();
+
+	if (pool.maxJobs <= 0) return newInstances;
+
+	while (pool.instanceInfo.size < pool.maxJobs) {
+		const info = new PoolInfo();
+		const instance = await pool.spawn();
+		pool.instanceInfo.set(instance, info);
+		info.active = true;
+		newInstances.add(instance);
+	}
+
+	return newInstances;
+};
+
+/**
+ * Reaps children if they've passed the maxRequest count.
+ * @param pool the pool object to work on
+ * @private
+ */
+const reap = (pool: Pool) => {
+	for (const [instance, info] of pool.instanceInfo) {
+		if (pool.maxRequests > 0 && info.requests >= pool.maxRequests) {
+			info.active = false;
+			pool.instanceInfo.delete(instance);
+			pool.reap(instance);
+			continue;
+		}
+	}
+};
+
+/**
+ * Handle fatal errors gracefully.
+ * @param pool the pool object to work on
+ * @param instance the php instance to clean up
+ * @param error the actual error that got us here
+ * @private
+ */
+const fatal = (pool: Pool, instance: instance, error: Error) => {
+	console.error(error);
+
+	if (instance && pool.instanceInfo.has(instance)) {
+		const info = pool.instanceInfo.get(instance);
+		info.active = false;
+		pool.instanceInfo.delete(instance);
+	}
+
+	return pool.fatal(instance, error);
+};
+
+/**
+ * Find the next available idle instance.
+ * @private
+ */
+const getIdleInstance = (pool) => {
```

Comment:
What's the type of `pool`?

Reply:
@adamziel https://github.com/WordPress/playground-tools/pull/110/files?diff=unified&w=1#diff-95251f05603f485be0377cd57ce90fe740821ca840c6053a115cb48c8725341bR91